File: CCserver/CarCare/BL/Log_Record.cs
using Microsoft.Extensions.Hosting;

namespace server.BL
{
    public class Log_Record
    {
        int logID;
        int currentKM;
        DateTime recordDate;
        DateTime warrantyExpirationDate;
        int cost;
        string notes;
        int mispar_mosah;
        int careID;
		string useremail;
		int licensePlate;
		string invoiceFileName;
		string carename;
		string garagename;

		public int LogID { get => logID; set => logID = value; }
		public int CurrentKM { get => currentKM; set => currentKM = value; }
		public DateTime RecordDate { get => recordDate; set => recordDate = value; }
		public DateTime WarrantyExpirationDate { get => warrantyExpirationDate; set => warrantyExpirationDate = value; }
		public int Cost { get => cost; set => cost = value; }
		public string Notes { get => notes; set => notes = value; }
		public int Mispar_mosah { get => mispar_mosah; set => mispar_mosah = value; }
		public int CareID { get => careID; set => careID = value; }
		public string Useremail { get => useremail; set => useremail = value; }
		public int LicensePlate { get => licensePlate; set => licensePlate = value; }
		public string InvoiceFileName { get => invoiceFileName; set => invoiceFileName = value; }
		public string Carename { get => carename; set => carename = value; }
		public string GarageName { get => garagename; set => garagename = value; }

		public Log_Record() { }

		public Log_Record(int logID, int currentKM, DateTime recordDate, DateTime warrantyExpirationDate, int cost, string notes, int mispar_mosah, int careID, string useremail, int licensePlate, string invoiceFileName, string carename,string garagename)
		{
			this.LogID = logID;
			this.CurrentKM = currentKM;
			this.RecordDate = recordDate;
			this.WarrantyExpirationDate = warrantyExpirationDate;
			this.Cost = cost;
			this.Notes = notes;
			this.Mispar_mosah = mispar_mosah;
			this.CareID = careID;
			this.Useremail = useremail;
			this.LicensePlate = licensePlate;
			this.InvoiceFileName = invoiceFileName;
			this.Carename = carename;
			this.GarageName = garagename;


		}

		public int InsertLog_Record()
        {
            DBservices dbs = new DBservices();

            return dbs.InsertLog_Record(this);
        }


		public int DeleteLog()
		{
			DBservices dbs = new DBservices();
			return dbs.DeleteLogRecord(this.LogID);
		}

		public List<Log_Record> GetCarUserLog(string email)
		{
			DBservices dbs = new DBservices();
			return dbs.GetCarUserLog(email);
		}

		public List<Log_Record> ReadLogRecord(int LogID)
		{
			DBservices dbs = new DBservices();
			return dbs.ReadLogRecord(LogID);
		}

		public int UpdateLog()
		{
			DBservices dbs = new DBservices();
			return dbs.UpdateLog(this);

		}

	}
}

File: server/routers/log_record.py
from fastapi import APIRouter, HTTPException
from database import get_connection
from models.schemas import LogRecordCreate, LogRecordUpdate

router = APIRouter(prefix="/api/logs", tags=["Log Records"])


def row_to_log(r) -> dict:
    return {
        "log_id": r[0], "current_km": r[1], "record_date": r[2],
        "cost": r[3], "notes": r[4],
        "garage_id": r[5], "care_id": r[6], "user_email": r[7],
        "license_plate": r[8], "invoice_file_name": r[9],
        "care_name": r[10] if len(r) > 10 else None,
        "garage_name": r[11] if len(r) > 11 else None,
    }


@router.get("/{email}")
def get_logs_by_email(email: str):
    try:
        conn = get_connection()
        cursor = conn.cursor()
        cursor.execute(
            """SELECT lr.LogID, lr.CurrentKM, lr.RecordDate,
                      lr.Cost, lr.Notes, lr.Mispar_mosah, lr.CareID, lr.UserEmail,
                      lr.LicensePlate, lr.InvoiceFileName, ct.CareName,
                      COALESCE(lr.GarageName, g.Shem_mosah) AS GarageName
               FROM Log_Record lr
               LEFT JOIN CareType ct ON lr.CareID = ct.CareID
               LEFT JOIN Garage g ON lr.Mispar_mosah = g.Id
               WHERE lr.UserEmail = ?""",
            email
        )
        return [row_to_log(r) for r in cursor.fetchall()]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        conn.close()


@router.get("/by-id/{log_id}")
def get_log_by_id(log_id: int):
    try:
        conn = get_connection()
        cursor = conn.cursor()
        cursor.execute(
            """SELECT lr.LogID, lr.CurrentKM, lr.RecordDate,
                      lr.Cost, lr.Notes, lr.Mispar_mosah, lr.CareID, lr.UserEmail,
                      lr.LicensePlate, lr.InvoiceFileName, ct.CareName,
                      COALESCE(lr.GarageName, g.Shem_mosah) AS GarageName
               FROM Log_Record lr
               LEFT JOIN CareType ct ON lr.CareID = ct.CareID
               LEFT JOIN Garage g ON lr.Mispar_mosah = g.Id
               WHERE lr.LogID = ?""",
            log_id
        )
        rows = cursor.fetchall()
        if not rows:
            raise HTTPException(status_code=404, detail="Log not found")
        return [row_to_log(r) for r in rows]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        conn.close()


@router.post("/", status_code=201)
def create_log(log: LogRecordCreate):
    try:
        conn = get_connection()
        cursor = conn.cursor()

        # Garages now come from the gov.il API, so their IDs won't exist in the
        # local Garage table. Only keep garage_id if it actually exists locally;
        # otherwise store NULL to avoid a foreign-key violation.
        garage_id = log.garage_id
        if garage_id is not None:
            cursor.execute("SELECT 1 FROM Garage WHERE Id = ?", garage_id)
            if cursor.fetchone() is None:
                garage_id = None

        cursor.execute(
            """INSERT INTO Log_Record
               (CurrentKM, RecordDate, Cost, Notes,
                Mispar_mosah, CareID, UserEmail, LicensePlate, InvoiceFileName, GarageName)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            log.current_km, log.record_date,
            log.cost, log.notes, garage_id, log.care_id,
            log.user_email, log.license_plate, log.invoice_file_name,
            log.garage_name
        )
        conn.commit()
        return {"message": "Log record created successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        conn.close()


@router.put("/{log_id}")
def update_log(log_id: int, log: LogRecordUpdate):
    try:
        conn = get_connection()
        cursor = conn.cursor()
        cursor.execute(
            """UPDATE Log_Record
               SET CurrentKM=?, RecordDate=?,
                   Cost=?, Notes=?, Mispar_mosah=?, CareID=?, InvoiceFileName=?
               WHERE LogID=?""",
            log.current_km, log.record_date,
            log.cost, log.notes, log.garage_id, log.care_id,
            log.invoice_file_name, log_id
        )
        conn.commit()
        if cursor.rowcount == 0:
            raise HTTPException(status_code=404, detail="Log not found")
        return {"message": "Log updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        conn.close()


@router.delete("/{log_id}")
def delete_log(log_id: int):
    try:
        conn = get_connection()
        cursor = conn.cursor()
        cursor.execute("DELETE FROM Log_Record WHERE LogID = ?", log_id)
        conn.commit()
        if cursor.rowcount == 0:
            raise HTTPException(status_code=404, detail="Log not found")
        return {"message": "Log deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        conn.close()
